### PR TITLE
Fix compilation with GCC and Clang

### PR DESCRIPTION
**Pull request by AngryLoki**
At this moment intel-extension-for-pytorch build fails both with GCC 13.2.1 and Clang 17/18/19.

GCC build fails with:
* failure to cast `c10:Half` to `_Float16`. While Clang-based compilers use [`operator float()` in `c10:Half`](https://github.com/pytorch/pytorch/blob/main/c10/util/Half.h#L382) and perform c10:Half->float->_Float16 conversion, gcc fails.
* Similar failure in `VecOps<__m512h>`, refuses to cast float to _Float16 automatically.

Clang fails with few issues:
* Compiler crashes with VLA+openmp combination, this was reported with a workaround at https://github.com/llvm/llvm-project/issues/75428
* `-Wl,-Bsymbolic-functions` should be a linker attribute; when applied to compiler, considered as "unused parameter" and fails due to `-Werror -Wall` combination.

Common issue:
* `extern int FLAGS_caffe2_log_level` should not be redeclared, otherwise build fails with different redefinition when buildsystem follows contradictory recommendation to enable GLog and GFlags from https://github.com/pytorch/pytorch/issues/44948

Warnings/misc:
* In Clang `-Wall` overrides some previous `-Wno-*` (e. g. `-Wno-unused-function -Wall`), so it should precede other flags
* Disabled few Clang warnings due to spam
* Added `f16` suffix in headers for f16 constants (spam in Clang/gcc)
* Fixed infinite recursion in `bool operator!=(const LlgaTensorDesc& desc)`

### CODE REVIEW

File: cmake/cpu/BuildFlags.cmake [hunk 1]
```diff
@@ -58,9 +59,14 @@ else(MSVC)
...
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-error=old-style-cast")
-  # Eigen fails to build with some versions, so convert this to a warning
-  # Details at http://eigen.tuxfamily.org/bz/show_bug.cgi?id=1459
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-ignored-attributes")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-pessimizing-move")
+
+  if (CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-vla-cxx-extension")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-deprecated-builtins")
```
Comment (jgong5) [inline cmake/cpu/BuildFlags.cmake:67]: can you comment on the build errors you found without these?

**Review by jgong5 — APPROVED** [review]
LGTM in general. I guess we need to have a validation to check if the changes still work fine with older compiler toolchains. cc @xuhancn to review and stamp on it.

File: cmake/cpu/BuildFlags.cmake (hunk 1 above)
Comment (jgong5) [reply cmake/cpu/BuildFlags.cmake:67]: Thanks. I guess we should fix most of these warnings instead of just ignoring them. @xuhancn can you take a look?

**Comment by xuhancn** [comment]
@jingxu10 will help on land this PR.

**Comment by jingxu10** [comment]
Hi @AngryLoki  there are several flake8 format issues. Could you follow the steps below to run the flake8 check at your side?
```
python -m pip install -r scripts/tools/setup/requirements-flake8.txt
python scripts/tools/setup/flake8.py
```

**Comment by AngryLoki** [comment]
@jingxu10 Python 3.11.8 and flake8==3.8.2 (current version) has no complains
<details><summary>Details</summary>
<p>

```
python scripts/tools/setup/flake8.py
All done! ✨ 🍰 ✨
1 file left unchanged.

status code:  0
All done! ✨ 🍰 ✨
156 files left unchanged.

status code:  0
All done! ✨ 🍰 ✨
3 files left unchanged.

status code:  0
All done! ✨ 🍰 ✨
103 files left unchanged.

status code:  0
Pass!
```
</p>
</details> 

Moreover, I don't change Python files in this this pull-request, so it should not break anything.

**Comment by jingxu10** [comment]
We have a clangformat check which failed in our internal CI system.
We are currently working on establishing the format check in the external CI system. I'll update you once it is online.

**Comment by xuhancn** [comment]
FYI: https://github.com/intel/intel-extension-for-pytorch/blob/main/CONTRIBUTING.md#code-style @AngryLoki

**Comment by jingxu10** [comment]
Hi, we have enabled a CI to run format verification. When you rebase your pr, the CI task should be triggered automatically.

**Comment by jingxu10** [comment]
Compilation failed.
[build_ipex_err.log](https://github.com/user-attachments/files/16263362/build_ipex_err.log)

**Comment by AngryLoki** [comment]
@jingxu10 , thank you for checking! 
Yes, `operator""f16` was added in gcc 13.1. And out of all options, only `_mm512_set1_ph(static_cast<_Float16>(1.0f))` produces no warnings:

```c++
#include <immintrin.h>

// note: clang 13 and gcc 11.4 don't support -mavx512fp16 at all

// works only since gcc 13.1 and clang 14.
// gcc 12.4 error: unable to find numeric literal operator 'operator""f16'
auto vec_zero1 = _mm512_set1_ph(1.0f16);

// gcc >= 13.1 warning: converting to '_Float16' from 'float' with greater conversion rank
// -Wno-narrowing removes warning, but it is not the best idea 
auto vec_zero2 = _mm512_set1_ph(1.0f);

// same as above: produces warning with gcc
auto vec_zero3 = _mm512_set1_ph((_Float16)1.0f);

// works everywhere, no warnings!
// (clang 13 and gcc 11.4 don't support -mavx512fp16)
auto vec_zero4 = _mm512_set1_ph(static_cast<_Float16>(1.0f));
```
(same in godbolt conformance view: https://godbolt.org/z/MdGbT7beT)

I've updated my pull-request with `static_cast` and rebased just in case.

**Comment by AngryLoki** [comment]
@DiweiSun , CI fails with

```
pip install lintrunner pip install lintrunner-adapters
ERROR: No matching distribution found for install
```

Could you update it? Thanks
```diff
-pip install lintrunner pip install lintrunner-adapters
+pip install lintrunner lintrunner-adapters
```
https://github.com/intel/intel-extension-for-pytorch/blob/5cc852eb850f8b42ac2d0fcca5709e6076e8042d/.github/workflows/format-check.yml#L32C11-L32C65

**Comment by DiweiSun** [comment]
> @DiweiSun , CI fails with
> 
> ```
> pip install lintrunner pip install lintrunner-adapters
> ERROR: No matching distribution found for install
> ```
> 
> Could you update it? Thanks
> 
> ```diff
> -pip install lintrunner pip install lintrunner-adapters
> +pip install lintrunner lintrunner-adapters
> ```
> 
> https://github.com/intel/intel-extension-for-pytorch/blob/5cc852eb850f8b42ac2d0fcca5709e6076e8042d/.github/workflows/format-check.yml#L32C11-L32C65

Fixed with https://github.com/intel/intel-extension-for-pytorch/commit/d12ebc5136cc98d699a730cf82e5fdc038162a47

**Comment by jingxu10** [comment]
merged to main branch

**Pull request closed without merging**